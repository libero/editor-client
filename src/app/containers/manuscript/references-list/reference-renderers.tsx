import React from 'react';
import { EditorState } from 'prosemirror-state';
import { get } from 'lodash';

import {
  BookReference,
  ConferenceReference,
  DataReference,
  JournalReference,
  PatentReference,
  PeriodicalReference,
  PrePrintReference,
  Reference,
  ReportReference,
  SoftwareReference,
  ThesisReference,
  WebReference
} from 'app/models/reference';
import { stringifyEditorState } from 'app/utils/view.utils';

const getAnnotatedText = (editorState: EditorState) => {
  const html = stringifyEditorState(editorState);
  return <span dangerouslySetInnerHTML={{ __html: html }} />;
};

export const renderJournalReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as JournalReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.articleTitle)}.{' '}
      <em>{getAnnotatedText(referenceInfo.source)}</em> {referenceInfo.inPress ? ' In Press.' : undefined}
      <strong>{referenceInfo.volume ? ` ${referenceInfo.volume}:` : undefined}</strong>
      {referenceInfo.firstPage && referenceInfo.lastPage
        ? ` p.${referenceInfo.firstPage}-${referenceInfo.lastPage}.`
        : undefined}
      {referenceInfo.elocationId ? ` ${referenceInfo.elocationId}.` : undefined}
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}
    </>
  );
};

export const renderBookReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as BookReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.chapterTitle)}.
      <em>{getAnnotatedText(referenceInfo.source)}</em> {referenceInfo.inPress ? ' In Press.' : undefined}
      {referenceInfo.edition ? ` ${referenceInfo.edition}.` : undefined}
      <strong>{referenceInfo.volume ? ` ${referenceInfo.volume}:` : undefined}</strong>
      {referenceInfo.firstPage && referenceInfo.lastPage
        ? ` p.${referenceInfo.firstPage}-${referenceInfo.lastPage}.`
        : undefined}
      {referenceInfo.elocationId ? ` ${referenceInfo.elocationId}.` : undefined}
      {referenceInfo.publisherName ? ` ${referenceInfo.publisherName}:` : undefined}
      {referenceInfo.publisherLocation ? ` ${referenceInfo.publisherLocation}.` : undefined}
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}
    </>
  );
};

export const renderConferenceReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as ConferenceReference;
  /*
  *   year: number;
  articleTitle: EditorState;
  conferenceName: EditorState;
  conferenceLocation: string;
  conferenceDate: string;
  volume: number;
  extLink: string;
  elocationId: string;
  doi: string;
  pmid: string;
  firstPage: number;
  lastPage: number;
  * */
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.articleTitle)}.
      <strong>{referenceInfo.volume ? ` ${referenceInfo.volume}:` : undefined}</strong>
      {referenceInfo.firstPage && referenceInfo.lastPage
        ? ` p.${referenceInfo.firstPage}-${referenceInfo.lastPage}.`
        : undefined}
      {referenceInfo.conferenceLocation ? ` ${referenceInfo.conferenceLocation}:` : undefined}
      {referenceInfo.conferenceName ? [getAnnotatedText(referenceInfo.conferenceName), '.'] : undefined}
      {referenceInfo.conferenceDate ? ` ${referenceInfo.conferenceDate}.` : undefined}
      {referenceInfo.elocationId ? ` ${referenceInfo.elocationId}.` : undefined}
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}
    </>
  );
};

export const renderThesisReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as ThesisReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.articleTitle)}.
      {referenceInfo.publisherName ? ` ${referenceInfo.publisherName}:` : undefined}
      {referenceInfo.publisherLocation ? ` ${referenceInfo.publisherLocation}.` : undefined}
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}{' '}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined}
    </>
  );
};

export const renderSoftwareReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as SoftwareReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.dataTitle)}.
      <em>{getAnnotatedText(referenceInfo.source)}</em>
      {referenceInfo.publisherName ? ` ${referenceInfo.publisherName}:` : undefined}
      {referenceInfo.publisherLocation ? ` ${referenceInfo.publisherLocation}.` : undefined}
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}
      {referenceInfo.version ? ` version: ${referenceInfo.version}. ` : undefined}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined}
    </>
  );
};

export const renderReportReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as ReportReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. <em>{getAnnotatedText(referenceInfo.source)}</em>
      {referenceInfo.patent ? ` ${referenceInfo.patent}.` : undefined}
      {referenceInfo.publisherName ? ` ${referenceInfo.publisherName}. ` : undefined}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined}
    </>
  );
};

export const renderPeriodicalReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as PeriodicalReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.date}. {getAnnotatedText(referenceInfo.articleTitle)}.{' '}
      <em>{getAnnotatedText(referenceInfo.source)}</em>
      <strong>{referenceInfo.volume ? ` ${referenceInfo.volume}:` : undefined}</strong>
      {referenceInfo.firstPage && referenceInfo.lastPage
        ? ` p.${referenceInfo.firstPage}-${referenceInfo.lastPage}. `
        : undefined}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined}
    </>
  );
};

export const renderPatentReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as PatentReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.articleTitle)}.{' '}
      <em>{getAnnotatedText(referenceInfo.source)}</em>
      {referenceInfo.publisherName ? ` ${referenceInfo.publisherName}.` : undefined}
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}{' '}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined}
    </>
  );
};

export const renderWebReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as WebReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.articleTitle)}.{' '}
      <em>{getAnnotatedText(referenceInfo.source)}</em>{' '}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined} {referenceInfo.dateInCitation}
    </>
  );
};

export const renderPreprintReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as PrePrintReference;
  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.articleTitle)}.{' '}
      <em>{getAnnotatedText(referenceInfo.source)}</em>
      {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.pmid ? ` pmid: ${referenceInfo.pmid}.` : undefined}{' '}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined}
    </>
  );
};

export const renderDataReference = (reference: Reference) => {
  const referenceInfo = reference.referenceInfo as DataReference;
  const specificUse = {
    generated: 'Generated',
    analyzed: 'Analyzed'
  }[referenceInfo.specificUse];

  const authors = getReferenceAuthors(reference);
  return (
    <>
      {authors}. {referenceInfo.year}. {getAnnotatedText(referenceInfo.dataTitle)}.{' '}
      <em>{getAnnotatedText(referenceInfo.source)}</em> {referenceInfo.doi ? ` doi: ${referenceInfo.doi}.` : undefined}
      {referenceInfo.accessionId ? ` pmid: ${referenceInfo.accessionId}.` : undefined}
      {referenceInfo.version ? ` version: ${referenceInfo.version}.` : undefined}{' '}
      {referenceInfo.extLink ? getExtLinkTag(referenceInfo.extLink) : undefined} {specificUse}
    </>
  );
};

const getReferenceAuthors = (reference: Reference) => {
  return reference.authors
    .map((author) => {
      return get(author, 'groupName', `${author['lastName']} ${author['firstName']}`);
    })
    .join(', ');
};

const getExtLinkTag = (url: string) => {
  return (
    <>
      url:{' '}
      <a target="_blank" rel="noopener noreferrer" href={url}>
        {url}
      </a>
    </>
  );
};
