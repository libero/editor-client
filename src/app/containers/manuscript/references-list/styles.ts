import { makeStyles } from '@material-ui/core/styles';

export const useReferencesListStyles = makeStyles((theme) => ({
  list: {
    marginLeft: theme.spacing(3),
    marginTop: 0,
    listStyleType: 'disc',
    '& li + li': {
      marginTop: theme.spacing(3)
    }
  }
}));

export const useReferencesListItemStyles = makeStyles((theme) => ({
  listItem: {
    paddingLeft: theme.spacing(1),
    display: 'flex',
    alignItems: 'flex-start'
  },
  content: {
    flexGrow: 1,
    lineHeight: '24px',
    paddingRight: theme.spacing(3)
  },
  editButton: {
    flexGrow: 0,
    padding: 6
  }
}));
