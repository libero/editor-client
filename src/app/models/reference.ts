import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';

import * as referenceInfoConfig from 'app/models/config/reference-info.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { getTextContentFromPath, makeSchemaFromConfig } from 'app/models/utils';

type ReferencePerson =
  | {
      firstName: string;
      lastName: string;
    }
  | {
      groupName: string;
    };

type ReferenceType =
  | 'journal'
  | 'periodical'
  | 'book'
  | 'report'
  | 'data'
  | 'web'
  | 'preprint'
  | 'software'
  | 'confproc'
  | 'thesis'
  | 'patent';

export interface JournalReference {
  year: number;
  source: EditorState;
  articleTitle: EditorState;
  volume: number;
  firstPage: string;
  lastPage: string;
  elocationId: string;
  inPress: boolean;
  doi: string;
  pmid: string;
}

export interface PeriodicalReference {
  date: string;
  source: EditorState;
  articleTitle: EditorState;
  volume: number;
  firstPage: string;
  lastPage: string;
  extLink: string;
}

export interface BookReference {
  year: number;
  chapterTitle: EditorState;
  edition: string;
  publisherLocation: string;
  publisherName: string;
  source: EditorState;
  volume: number;
  editors: ReferencePerson[];
  firstPage: string;
  lastPage: string;
  doi: string;
  pmid: string;
  inPress: boolean;
  elocationId: string;
}

export interface ReportReference {
  year: number;
  source: EditorState;
  publisherName: string;
  doi: string;
  extLink: string;
}

export interface PatentReference {
  year: number;
  source: EditorState;
  articleTitle: EditorState;
  publisherName: string;
  doi: string;
  pmid: string;
  patent: string;
  extLink: string;
}

export interface DataReference {
  year: number;
  dataTitle: EditorState;
  source: EditorState;
  doi: string;
  version: string;
  extLink: string;
  accessionId: string;
  accessionUrl: string;
  specificUse: string;
}

export interface WebReference {
  year: number;
  source: EditorState;
  articleTitle: EditorState;
  extLink: string;
  dateInCitation: string;
}

export interface PrePrintReference {
  year: number;
  articleTitle: EditorState;
  source: EditorState;
  extLink: string;
  doi: string;
  pmid: string;
}

export interface ThesisReference {
  year: number;
  articleTitle: EditorState;
  publisherName: string;
  publisherLocation: string;
  extLink: string;
  doi: string;
  pmid: string;
}

export interface SoftwareReference {
  year: number;
  source: EditorState;
  dataTitle: EditorState;
  version: string;
  publisherName: string;
  publisherLocation: string;
  extLink: string;
  doi: string;
  pmid: string;
}

export interface ConferenceReference {
  year: number;
  articleTitle: EditorState;
  conferenceName: EditorState;
  conferenceLocation: string;
  conferenceDate: string;
  volume: number;
  extLink: string;
  elocationId: string;
  doi: string;
  pmid: string;
  firstPage: string;
  lastPage: string;
}

export interface Reference {
  id: string;
  authors: Array<ReferencePerson>;
  type: ReferenceType;
  referenceInfo:
    | JournalReference
    | BookReference
    | ConferenceReference
    | DataReference
    | PeriodicalReference
    | PrePrintReference
    | ReportReference
    | PatentReference
    | SoftwareReference
    | WebReference
    | ThesisReference;
}

export function createReference(xmlId: string, referenceXml: Element): Reference {
  const reference: Reference = {
    id: xmlId,
    authors: [
      ...createReferencePersonList(referenceXml, 'author'),
      ...createReferencePersonList(referenceXml, 'inventor')
    ],
    type: referenceXml.getAttribute('publication-type') as ReferenceType,
    referenceInfo: null
  };

  switch (reference.type) {
    case 'journal':
      reference.referenceInfo = createJournalReference(referenceXml);
      break;
    case 'book':
      reference.referenceInfo = createBookReference(referenceXml);
      break;
    case 'periodical':
      reference.referenceInfo = createPeriodicalReference(referenceXml);
      break;
    case 'report':
      reference.referenceInfo = createReportReference(referenceXml);
      break;
    case 'data':
      reference.referenceInfo = createDataReference(referenceXml);
      break;
    case 'web':
      reference.referenceInfo = createWebReference(referenceXml);
      break;
    case 'preprint':
      reference.referenceInfo = createPrePrintReference(referenceXml);
      break;
    case 'software':
      reference.referenceInfo = createSoftwareReference(referenceXml);
      break;
    case 'confproc':
      reference.referenceInfo = createConferenceReference(referenceXml);
      break;
    case 'thesis':
      reference.referenceInfo = createThesisReference(referenceXml);
      break;
    case 'patent':
      reference.referenceInfo = createPatentReference(referenceXml);
      break;
  }

  return reference;
}

function createReferencePersonList(referenceXml: Element, groupType: string): ReferencePerson[] {
  const contributors = referenceXml.querySelector(`person-group[person-group-type=${groupType}]`);
  if (!contributors) {
    return [];
  }

  return Array.from(contributors.children).map((contributorXml) => {
    if (contributorXml.nodeName === 'name') {
      return {
        firstName: getTextContentFromPath(referenceXml, 'given-names') || '',
        lastName: getTextContentFromPath(contributorXml, 'surname') || ''
      };
    } else {
      return {
        groupName: contributorXml.textContent
      };
    }
  });
}

function createJournalReference(referenceXml: Element): JournalReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || '',
    elocationId: getTextContentFromPath(referenceXml, 'elocation-id') || '',
    firstPage: getTextContentFromPath(referenceXml, 'fpage'),
    lastPage: getTextContentFromPath(referenceXml, 'lpage'),
    inPress: getTextContentFromPath(referenceXml, 'comment') === 'In press',
    volume: parseInt(getTextContentFromPath(referenceXml, 'volume') || '0')
  };
}

function createPatentReference(referenceXml: Element): PatentReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    publisherName: getTextContentFromPath(referenceXml, 'publisher-name') || '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || '',
    patent: getTextContentFromPath(referenceXml, 'patent') || '',
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || ''
  };
}

function createPeriodicalReference(referenceXml: Element): PeriodicalReference {
  const day = getTextContentFromPath(referenceXml, 'string-date > day');
  const month = getTextContentFromPath(referenceXml, 'string-date > month');
  const year = getTextContentFromPath(referenceXml, 'string-date > year');

  return {
    date: `${year}-${month}-${day}`,
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    firstPage: getTextContentFromPath(referenceXml, 'fpage'),
    lastPage: getTextContentFromPath(referenceXml, 'lpage'),
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || '',
    volume: parseInt(getTextContentFromPath(referenceXml, 'volume') || '0')
  };
}

function createReportReference(referenceXml: Element): ReportReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    publisherName: getTextContentFromPath(referenceXml, 'publisher-name') || '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || '',
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || ''
  };
}

function createSoftwareReference(referenceXml: Element): SoftwareReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    dataTitle: createReferenceAnnotatedValue(referenceXml.querySelector('data-title')),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    version: getTextContentFromPath(referenceXml, 'version') || '',
    publisherLocation: getTextContentFromPath(referenceXml, 'publisher-loc') || '',
    publisherName: getTextContentFromPath(referenceXml, 'publisher-name') || '',
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || ''
  };
}

function createThesisReference(referenceXml: Element): ThesisReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    publisherLocation: getTextContentFromPath(referenceXml, 'publisher-loc') || '',
    publisherName: getTextContentFromPath(referenceXml, 'publisher-name') || '',
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || ''
  };
}

function createWebReference(referenceXml: Element): WebReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || '',
    dateInCitation: getTextContentFromPath(referenceXml, 'date-in-citation') || ''
  };
}
function createConferenceReference(referenceXml: Element): ConferenceReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    conferenceName: createReferenceAnnotatedValue(referenceXml.querySelector('conf-name')),
    conferenceLocation: getTextContentFromPath(referenceXml, 'conf-loc') || '',
    conferenceDate: getTextContentFromPath(referenceXml, 'conf-date') || '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || '',
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || '',
    firstPage: getTextContentFromPath(referenceXml, 'fpage'),
    lastPage: getTextContentFromPath(referenceXml, 'lpage'),
    elocationId: getTextContentFromPath(referenceXml, 'elocation-id') || '',
    volume: parseInt(getTextContentFromPath(referenceXml, 'volume') || '0')
  };
}

function createPrePrintReference(referenceXml: Element): PrePrintReference {
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    articleTitle: createReferenceAnnotatedValue(referenceXml.querySelector('article-title')),
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || ''
  };
}

function createDataReference(referenceXml: Element): DataReference {
  const accessionEl = referenceXml.querySelector('pub-id[pub-id-type="accession"]');
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    dataTitle: createReferenceAnnotatedValue(referenceXml.querySelector('data-title')),
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    accessionId: accessionEl ? accessionEl.textContent : '',
    accessionUrl: accessionEl ? accessionEl.getAttribute('href') : '',
    extLink: getTextContentFromPath(referenceXml, 'ext-link') || '',
    version: getTextContentFromPath(referenceXml, 'version') || '',
    specificUse: referenceXml.getAttribute('specific-use')
  };
}

function createBookReference(referenceXml: Element): BookReference {
  const editors: ReferencePerson[] = createReferencePersonList(referenceXml, 'editor');
  return {
    year: parseInt(referenceXml.querySelector('year').textContent),
    source: createReferenceAnnotatedValue(referenceXml.querySelector('source')),
    chapterTitle: createReferenceAnnotatedValue(referenceXml.querySelector('chapter-title')),
    publisherLocation: getTextContentFromPath(referenceXml, 'publisher-loc') || '',
    editors: editors,
    publisherName: getTextContentFromPath(referenceXml, 'publisher-name') || '',
    edition: getTextContentFromPath(referenceXml, 'edition') || '',
    doi: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="doi"]') || '',
    pmid: getTextContentFromPath(referenceXml, 'pub-id[pub-id-type="pmid"]') || '',
    elocationId: getTextContentFromPath(referenceXml, 'elocation-id') || '',
    firstPage: getTextContentFromPath(referenceXml, 'fpage'),
    lastPage: getTextContentFromPath(referenceXml, 'lpage'),
    inPress: getTextContentFromPath(referenceXml, 'comment') === 'In press',
    volume: parseInt(getTextContentFromPath(referenceXml, 'volume') || '0')
  };
}

export function createReferenceAnnotatedValue(content: Node): EditorState {
  const schema = makeSchemaFromConfig(
    referenceInfoConfig.topNode,
    referenceInfoConfig.nodes,
    referenceInfoConfig.marks
  );

  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), gapCursor(), dropCursor()]
  });
}
