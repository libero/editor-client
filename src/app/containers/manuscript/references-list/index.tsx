import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { SectionContainer } from 'app/components/section-container';
import { ActionButton } from 'app/components/action-button';
import { getReferences } from 'app/selectors/manuscript.selectors';
import { Reference } from 'app/models/reference';
import { useReferencesListItemStyles, useReferencesListStyles } from 'app/containers/manuscript/references-list/styles';
import {
  renderBookReference,
  renderConferenceReference,
  renderDataReference,
  renderJournalReference,
  renderPatentReference,
  renderPeriodicalReference,
  renderPreprintReference,
  renderReportReference,
  renderSoftwareReference,
  renderThesisReference,
  renderWebReference
} from 'app/containers/manuscript/references-list/reference-renderers';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ConnectedReferenceFormDialog } from 'app/containers/reference-form-dialog/connected-reference-form-dialog';

interface ReferenceItemProps {
  onEditCallback: (reference: Reference) => void;
  reference: Reference;
}

const renderReferenceContent = (reference: Reference) => {
  switch (reference.type) {
    case 'journal':
      return renderJournalReference(reference);
    case 'book':
      return renderBookReference(reference);
    case 'report':
      return renderReportReference(reference);
    case 'periodical':
      return renderPeriodicalReference(reference);
    case 'patent':
      return renderPatentReference(reference);
    case 'web':
      return renderWebReference(reference);
    case 'preprint':
      return renderPreprintReference(reference);
    case 'data':
      return renderDataReference(reference);
    case 'confproc':
      return renderConferenceReference(reference);
    case 'software':
      return renderSoftwareReference(reference);
    case 'thesis':
      return renderThesisReference(reference);
  }
};

const ReferenceItem: React.FC<ReferenceItemProps> = ({ onEditCallback, reference }) => {
  const classes = useReferencesListItemStyles();

  return (
    <li>
      <section className={classes.listItem}>
        <div className={classes.content}>{renderReferenceContent(reference)}</div>
        <IconButton onClick={onEditCallback.bind(null, reference)} classes={{ root: classes.editButton }}>
          <EditIcon fontSize="small" />
        </IconButton>
      </section>
    </li>
  );
};

export const ReferenceList: React.FC<{}> = () => {
  const references = useSelector(getReferences);
  const classes = useReferencesListStyles();
  const dispatch = useDispatch();

  const handleAddReference = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ConnectedReferenceFormDialog,
        title: 'Reference'
      })
    );
  }, [dispatch]);

  const handleEditReference = useCallback(
    (reference: Reference) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: ConnectedReferenceFormDialog,
          title: 'Reference',
          props: { reference }
        })
      );
    },
    [dispatch]
  );

  return (
    <section>
      <SectionContainer label="References">
        <ul className={classes.list}>
          {references.map((reference) => (
            <ReferenceItem key={reference.id} reference={reference} onEditCallback={handleEditReference} />
          ))}
        </ul>
      </SectionContainer>
      <ActionButton variant="addEntity" title="Reference" onClick={handleAddReference} />
    </section>
  );
};
