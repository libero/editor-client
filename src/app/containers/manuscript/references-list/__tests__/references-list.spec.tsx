import React from 'react';
import configureMockStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { EditorState } from 'prosemirror-state';
import { Provider } from 'react-redux';

import { ReferenceList } from 'app/containers/manuscript/references-list/index';
import { givenState } from 'app/test-utils/reducer-test-helpers';
import { createReferenceAnnotatedValue, Reference } from 'app/models/reference';

const ReferenceData = [
  {
    id: 'bib1',
    authors: [
      {
        firstName: 'V',
        lastName: 'Berk'
      }
    ],
    type: 'journal',
    referenceInfo: {
      year: 2012,
      source: stringToEditorState('Science'),
      articleTitle: stringToEditorState(
        'Molecular architecture and assembly principles of <italic>Vibrio cholerae</italic> biofilms'
      ),
      doi: '',
      pmid: '',
      elocationId: '',
      firstPage: '236',
      lastPage: '239',
      inPress: false,
      volume: 337
    }
  },
  {
    id: 'bib2',
    authors: [
      {
        firstName: 'PK',
        lastName: 'Feyerabend'
      }
    ],
    type: 'book',
    referenceInfo: {
      year: 2010,
      source: stringToEditorState('The skull and brain'),
      chapterTitle: stringToEditorState('The skull and brain'),
      publisherLocation: 'London',
      publisherName: 'Verso',
      edition: '4th Edition',
      editors: [],
      doi: '',
      pmid: '',
      elocationId: '',
      firstPage: 0,
      lastPage: 0,
      inPress: false,
      volume: 0
    }
  }
] as Reference[];

jest.mock('@material-ui/core', () => ({
  Button: ({ label }) => <div data-cmp="Button">{label}</div>,
  IconButton: () => <div data-cmp="IconButton"></div>
}));

describe('References List', () => {
  const mockStore = configureMockStore([]);

  it('renders the component', () => {
    const store = mockStore({ manuscript: givenState({ references: ReferenceData }) });

    const wrapper = create(
      <Provider store={store}>
        <ReferenceList />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});

function stringToEditorState(xmlContent: string): EditorState {
  const el = document.createElement('div');
  el.innerHTML = xmlContent;
  return createReferenceAnnotatedValue(el);
}
