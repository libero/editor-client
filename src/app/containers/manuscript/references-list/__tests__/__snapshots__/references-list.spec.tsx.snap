// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`References List renders the component 1`] = `
<section>
  <div
    className="makeStyles-root-2 makeStyles-plainVariant-4"
  >
    <label
      className="makeStyles-label-5"
    >
      References
    </label>
    <ul
      className="makeStyles-list-1"
    >
      <li>
        <section
          className="makeStyles-listItem-7"
        >
          <div
            className="makeStyles-content-8"
          >
            Berk V
            . 
            2012
            . 
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "Molecular architecture and assembly principles of <em>Vibrio cholerae</em> biofilms",
                }
              }
            />
            . 
            <em>
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Science",
                  }
                }
              />
              
            </em>
             
            <strong>
               337:
            </strong>
             p.236-239.
          </div>
          <div
            data-cmp="IconButton"
          />
        </section>
      </li>
      <li>
        <section
          className="makeStyles-listItem-7"
        >
          <div
            className="makeStyles-content-8"
          >
            Feyerabend PK
            . 
            2010
            . 
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "The skull and brain",
                }
              }
            />
            . 
            <em>
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "The skull and brain",
                  }
                }
              />
               
            </em>
             
             4th Edition.
            <strong />
             Verso:
             London.
          </div>
          <div
            data-cmp="IconButton"
          />
        </section>
      </li>
    </ul>
  </div>
  <div
    data-cmp="Button"
  />
</section>
`;
